### PR TITLE
go/common/crypto/signature/signers/plugin: Rewrite

### PR DESCRIPTION
**Pull request by Yawning**
The Go runtime library's `plugin` is a steaming pile of shit.  Switch
to something that lets people write plugins that live in a separate
module, that import sub-packages from the module that is loading the
plugin.

### CODE REVIEW

File: go/oasis-node/main.go [hunk 1]
```diff
@@ -2,9 +2,14 @@
 package main
 
 import (
+	"github.com/hashicorp/go-plugin"
+
 	"github.com/oasisprotocol/oasis-core/go/oasis-node/cmd"
 )
 
 func main() {
+	// If we use go-plugin, we are supposed to clean clients up.
```
Comment (kostko) [inline go/oasis-node/main.go:11]: Probably the same in remote signer?

Comment (Yawning) [reply go/oasis-node/main.go:11]: And in the tests yeah.  I'll deal with that after I unbreak the tests.

File: go/common/crypto/signature/signers/plugin/rpc.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
+)
+
+var _ Signer = (*rpcClient)(nil)
+
+type rpcServer struct {
+	impl Signer
+}
+
+// XXX: Do the arg types need to be public?
```
Comment (kostko) [inline go/common/crypto/signature/signers/plugin/rpc.go]: Do they need to be? All the encoding/decoding is done in this package.

File: go/common/crypto/signature/signers/plugin/plugin.go [hunk 3]
```diff
@@ -3,33 +3,68 @@ package plugin
...
+	return plugin.HandshakeConfig{
+		ProtocolVersion:  1,
+		MagicCookieKey:   "OASIS_CORE_SIGNER_PLUGIN",
+		MagicCookieValue: name,
+	}
+}
 
 // FactoryConfig is the plugin factory configuration.
 type FactoryConfig struct {
+	// Name is the expected human readable name of the plugin (eg: "ledger", "memory")
```
Comment (kostko) [inline go/common/crypto/signature/signers/plugin/plugin.go:44]: Why does the name need to be specified here? Shouldn't it be provided by the plugin?

File: go/common/crypto/signature/signers/plugin/rpc.go (hunk 2 above)
Comment (Yawning) [reply go/common/crypto/signature/signers/plugin/rpc.go]: Yes.

File: go/common/crypto/signature/signers/plugin/plugin.go (hunk 3 above)
Comment (Yawning) [reply go/common/crypto/signature/signers/plugin/plugin.go:44]: I could remove it entirely instead and just have every plugin named `signer` as far as the transport goes.  Figured I'd make users explicitly know what plugin the random binary they're passing to the node actually is.

**Pull request merged**